[Bug] Compile the ROCM backend，compile failed

-- Found TVM_LLVM_VERSION=200
-- Found TVM_LLVM_HAS_AARCH64_TARGET=0
-- Build with LLVM
-- Set TVM_LLVM_VERSION=200
-- Build with contrib.random
-- Build with contrib.sort
-- Git found: /usr/bin/git
-- Found TVM_GIT_COMMIT_HASH=87f478ed30a9b273e27f435d2fbd010dcf3b7848
-- Found TVM_GIT_COMMIT_TIME=2026-01-27 13:55:36 +0800
-- Detected target triple: x86_64-linux-gnu
-- Setting C++ macro TVM_FFI_USE_LIBBACKTRACE - 1
-- Setting C++ macro TVM_FFI_BACKTRACE_ON_SEGFAULT - 1
-- Building with TVM Map...
-- Build with thread support...
-- CLANG_VERSION 20.0
-- Setting enhanced clang warning flags
-- Added "-fuse-ld=lld" to linker flags
-- Configuring done (3.7s)
-- Generating done (0.2s)

/home/richard/data/tvm-build/tvm/src/target/llvm/codegen_llvm.cc:2284:29: error: no matching member function for call to 'insertDeclare'
 2284 |     dbg_info_->di_builder_->insertDeclare(
      |     ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/data/richard/rocm-all-libs-build/rocm-install/include/llvm/IR/DIBuilder.h:1018:16: note: candidate function not viable: no known conversion from 'llvm::BasicBlock::iterator' (aka 'ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, false>') to 'BasicBlock *' for 5th argument
 1018 |     DbgInstPtr insertDeclare(llvm::Value *Storage, DILocalVariable *VarInfo,
      |                ^
 1019 |                              DIExpression *Expr, const DILocation *DL,
 1020 |                              BasicBlock *InsertAtEnd);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~
/data/richard/rocm-all-libs-build/rocm-install/include/llvm/IR/DIBuilder.h:1047:16: note: candidate function not viable: no known conversion from 'llvm::BasicBlock::iterator' (aka 'ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, false>') to 'Instruction *' for 5th argument
 1047 |     DbgInstPtr insertDeclare(llvm::Value *Storage, DILocalVariable *VarInfo,
      |                ^
 1048 |                              DIExpression *Expr, const DILocation *DL,
 1049 |                              Instruction *InsertBefore);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~
/data/richard/rocm-all-libs-build/rocm-install/include/llvm/IR/DIBuilder.h:98:16: note: candidate function not viable: requires 6 arguments, but 5 were provided
   98 |     DbgInstPtr insertDeclare(llvm::Value *Storage, DILocalVariable *VarInfo,
      |                ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   99 |                              DIExpression *Expr, const DILocation *DL,
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  100 |                              BasicBlock *InsertBB, Instruction *InsertBefore);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/richard/data/tvm-build/tvm/src/target/llvm/codegen_llvm.cc:2328:29: error: no matching member function for call to 'insertDeclare'
 2328 |     dbg_info_->di_builder_->insertDeclare(
      |     ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/data/richard/rocm-all-libs-build/rocm-install/include/llvm/IR/DIBuilder.h:1018:16: note: candidate function not viable: no known conversion from 'llvm::BasicBlock::iterator' (aka 'ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, false>') to 'BasicBlock *' for 5th argument
 1018 |     DbgInstPtr insertDeclare(llvm::Value *Storage, DILocalVariable *VarInfo,
      |                ^
 1019 |                              DIExpression *Expr, const DILocation *DL,
 1020 |                              BasicBlock *InsertAtEnd);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~
/data/richard/rocm-all-libs-build/rocm-install/include/llvm/IR/DIBuilder.h:1047:16: note: candidate function not viable: no known conversion from 'llvm::BasicBlock::iterator' (aka 'ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, false>') to 'Instruction *' for 5th argument
 1047 |     DbgInstPtr insertDeclare(llvm::Value *Storage, DILocalVariable *VarInfo,
      |                ^
 1048 |                              DIExpression *Expr, const DILocation *DL,
 1049 |                              Instruction *InsertBefore);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~
/data/richard/rocm-all-libs-build/rocm-install/include/llvm/IR/DIBuilder.h:98:16: note: candidate function not viable: requires 6 arguments, but 5 were provided
   98 |     DbgInstPtr insertDeclare(llvm::Value *Storage, DILocalVariable *VarInfo,
      |                ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   99 |                              DIExpression *Expr, const DILocation *DL,
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  100 |                              BasicBlock *InsertBB, Instruction *InsertBefore);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated when compiling for gfx1102.

> Hi [@WsqRichards](https://github.com/WsqRichards), could you please update your codebase to the latest main? This issue is expected to have been fixed by [#18706](https://github.com/apache/tvm/pull/18706), which was merged yesterday. The error message you shared indicates that your codebase is not the latest. For example, your codebase is
> 
> ```
> /home/richard/data/tvm-build/tvm/src/target/llvm/codegen_llvm.cc:2328:29: error: no matching member function for call to 'insertDeclare'
> 2328 | dbg_info_->di_builder_->insertDeclare(
> | ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
> ```
> 
> In the latest main, the line 2328 is
> 
> [tvm/src/target/llvm/codegen_llvm.cc](https://github.com/apache/tvm/blob/efdf2b2dbb674d990e32b881c3de0e62986c9407/src/target/llvm/codegen_llvm.cc#L2326-L2330)
> 
> Lines 2326 to 2330 in [efdf2b2](/apache/tvm/commit/efdf2b2dbb674d990e32b881c3de0e62986c9407)
> 
>  #if TVM_LLVM_VERSION >= 150 
>    // LLVM 15+ requires dbg_declare to reference pointer or integer types only. 
>    // For non-pointer types (floats, vectors), use dbg_value instead to track 
>    // the SSA value directly rather than a memory location. 
>    if (!llvm_value->getType()->isPointerTy()) {

The compilation still failed. The compilation log is as follows
[build.log](https://github.com/user-attachments/files/25060220/build.log)

--- a/cmake/utils/FindLLVM.cmake
+++ b/cmake/utils/FindLLVM.cmake
@@ -238,15 +238,25 @@ macro(find_llvm use_llvm)
   endif()
   message(STATUS "Found TVM_LLVM_HAS_AARCH64_TARGET=" ${TVM_LLVM_HAS_AARCH64_TARGET})
 
-  # Detect whether DIBuilder::insertDeclare accepts BasicBlock::iterator
-  # as the insertion point (upstream LLVM 20+) vs Instruction* (pre-LLVM 20
-  # and ROCm-bundled LLVM 20, which reports version 200 but lacks this overload).
+  # Detect whether DIBuilder insertion APIs (insertDeclare,
+  # insertDbgValueIntrinsic) accept BasicBlock::iterator as the insertion point
+  # (upstream LLVM 20+) vs Instruction* (pre-LLVM 20 and ROCm-bundled LLVM 20,
+  # which reports version 200 but retains the legacy Instruction* overload).
   if (${TVM_LLVM_VERSION} GREATER_EQUAL 200)
     include(CheckCXXSourceCompiles)
     include(CMakePushCheckState)
     cmake_push_check_state(RESET)
     set(CMAKE_REQUIRED_INCLUDES ${LLVM_INCLUDE_DIRS})
-    set(CMAKE_REQUIRED_FLAGS "-std=c++17 -fno-rtti")
+    # Only force -std= when the outer project hasn't already selected a
+    # standard; signature-only probe does not need -fno-rtti. Use the
+    # compiler-appropriate form so the probe works under MSVC as well.
+    if(NOT CMAKE_CXX_STANDARD)
+      if(MSVC)
+        set(CMAKE_REQUIRED_FLAGS "/std:c++17")
+      else()
+        set(CMAKE_REQUIRED_FLAGS "-std=c++17")
+      endif()
+    endif()
     check_cxx_source_compiles("
       #include <llvm/IR/DIBuilder.h>
       #include <llvm/IR/Instructions.h>
@@ -254,14 +264,15 @@ macro(find_llvm use_llvm)
              llvm::DIExpression* e, const llvm::DILocation* dl,
              llvm::Instruction* i) {
         b->insertDeclare(v, var, e, dl, llvm::BasicBlock::iterator(i));
+        b->insertDbgValueIntrinsic(v, var, e, dl, llvm::BasicBlock::iterator(i));
       }
-    " TVM_LLVM_INSERTDECLARE_USES_ITERATOR)
+    " TVM_LLVM_DIBUILDER_USES_ITERATOR)
     cmake_pop_check_state()
-    if(TVM_LLVM_INSERTDECLARE_USES_ITERATOR)
-      add_definitions(-DTVM_LLVM_INSERTDECLARE_USES_ITERATOR=1)
-      message(STATUS "LLVM DIBuilder::insertDeclare uses BasicBlock::iterator")
+    if(TVM_LLVM_DIBUILDER_USES_ITERATOR)
+      add_definitions(-DTVM_LLVM_DIBUILDER_USES_ITERATOR=1)
+      message(STATUS "LLVM DIBuilder insertion APIs use BasicBlock::iterator")
     else()
-      message(STATUS "LLVM DIBuilder::insertDeclare uses Instruction* (ROCm or older LLVM 20)")
+      message(STATUS "LLVM DIBuilder insertion APIs use Instruction* (ROCm or older LLVM 20)")
     endif()
   endif()
 endmacro(find_llvm)

--- a/cmake/utils/FindLLVM.cmake
+++ b/cmake/utils/FindLLVM.cmake
@@ -237,4 +237,31 @@ macro(find_llvm use_llvm)
     message(FATAL_ERROR "TVM requires LLVM 15.0 or higher.")
   endif()
   message(STATUS "Found TVM_LLVM_HAS_AARCH64_TARGET=" ${TVM_LLVM_HAS_AARCH64_TARGET})
+
+  # Detect whether DIBuilder::insertDeclare accepts BasicBlock::iterator
+  # as the insertion point (upstream LLVM 20+) vs Instruction* (pre-LLVM 20
+  # and ROCm-bundled LLVM 20, which reports version 200 but lacks this overload).
+  if (${TVM_LLVM_VERSION} GREATER_EQUAL 200)
+    include(CheckCXXSourceCompiles)
+    include(CMakePushCheckState)
+    cmake_push_check_state(RESET)
+    set(CMAKE_REQUIRED_INCLUDES ${LLVM_INCLUDE_DIRS})
+    set(CMAKE_REQUIRED_FLAGS "-std=c++17 -fno-rtti")
+    check_cxx_source_compiles("
+      #include <llvm/IR/DIBuilder.h>
+      #include <llvm/IR/Instructions.h>
+      void f(llvm::DIBuilder* b, llvm::Value* v, llvm::DILocalVariable* var,
+             llvm::DIExpression* e, const llvm::DILocation* dl,
+             llvm::Instruction* i) {
+        b->insertDeclare(v, var, e, dl, llvm::BasicBlock::iterator(i));
+      }
+    " TVM_LLVM_INSERTDECLARE_USES_ITERATOR)
+    cmake_pop_check_state()
+    if(TVM_LLVM_INSERTDECLARE_USES_ITERATOR)
+      add_definitions(-DTVM_LLVM_INSERTDECLARE_USES_ITERATOR=1)
+      message(STATUS "LLVM DIBuilder::insertDeclare uses BasicBlock::iterator")
+    else()
+      message(STATUS "LLVM DIBuilder::insertDeclare uses Instruction* (ROCm or older LLVM 20)")
+    endif()
+  endif()
 endmacro(find_llvm)

--- a/src/target/llvm/codegen_llvm.cc
+++ b/src/target/llvm/codegen_llvm.cc
@@ -2147,7 +2147,7 @@ void CodeGenLLVM::AddDebugInformation(llvm::Function* f_llvm,
 
     auto* store = builder.CreateStore(iter_param, paramAlloca);
     auto* di_loc = llvm::DILocation::get(*ctx, 0, 0, di_subprogram_);
-#if TVM_LLVM_INSERTDECLARE_USES_ITERATOR
+#if TVM_LLVM_DIBUILDER_USES_ITERATOR
     dbg_info_->di_builder_->insertDeclare(
         paramAlloca, param, dbg_info_->di_builder_->createExpression(), llvm::DebugLoc(di_loc),
         llvm::BasicBlock::iterator(store));
@@ -2195,8 +2195,8 @@ void CodeGenLLVM::AddDebugInformation(llvm::Value* llvm_value, const Var& tir_va
     if (insert_before) {
       // Upstream LLVM 20+ changed insertDbgValueIntrinsic to take
       // BasicBlock::iterator; ROCm-bundled LLVM 20 retains Instruction*.
-      // TVM_LLVM_INSERTDECLARE_USES_ITERATOR is set by CMake feature detection.
-#if TVM_LLVM_INSERTDECLARE_USES_ITERATOR
+      // TVM_LLVM_DIBUILDER_USES_ITERATOR is set by CMake feature detection.
+#if TVM_LLVM_DIBUILDER_USES_ITERATOR
       dbg_info_->di_builder_->insertDbgValueIntrinsic(
           llvm_value, local_var, dbg_info_->di_builder_->createExpression(), llvm::DebugLoc(di_loc),
           llvm::BasicBlock::iterator(insert_before));
@@ -2214,7 +2214,7 @@ void CodeGenLLVM::AddDebugInformation(llvm::Value* llvm_value, const Var& tir_va
   }
 
   if (insert_before) {
-#if TVM_LLVM_INSERTDECLARE_USES_ITERATOR
+#if TVM_LLVM_DIBUILDER_USES_ITERATOR
     dbg_info_->di_builder_->insertDeclare(
         llvm_value, local_var, dbg_info_->di_builder_->createExpression(), llvm::DebugLoc(di_loc),
         llvm::BasicBlock::iterator(insert_before));

--- a/src/target/llvm/codegen_llvm.cc
+++ b/src/target/llvm/codegen_llvm.cc
@@ -2147,7 +2147,7 @@ void CodeGenLLVM::AddDebugInformation(llvm::Function* f_llvm,
 
     auto* store = builder.CreateStore(iter_param, paramAlloca);
     auto* di_loc = llvm::DILocation::get(*ctx, 0, 0, di_subprogram_);
-#if TVM_LLVM_VERSION >= 200
+#if TVM_LLVM_INSERTDECLARE_USES_ITERATOR
     dbg_info_->di_builder_->insertDeclare(
         paramAlloca, param, dbg_info_->di_builder_->createExpression(), llvm::DebugLoc(di_loc),
         llvm::BasicBlock::iterator(store));
@@ -2193,9 +2193,10 @@ void CodeGenLLVM::AddDebugInformation(llvm::Value* llvm_value, const Var& tir_va
   // the SSA value directly rather than a memory location.
   if (!llvm_value->getType()->isPointerTy()) {
     if (insert_before) {
-      // LLVM 20+ changed insertDbgValueIntrinsic to take BasicBlock::iterator
-      // instead of Instruction* for the insertion point.
-#if TVM_LLVM_VERSION >= 200
+      // Upstream LLVM 20+ changed insertDbgValueIntrinsic to take
+      // BasicBlock::iterator; ROCm-bundled LLVM 20 retains Instruction*.
+      // TVM_LLVM_INSERTDECLARE_USES_ITERATOR is set by CMake feature detection.
+#if TVM_LLVM_INSERTDECLARE_USES_ITERATOR
       dbg_info_->di_builder_->insertDbgValueIntrinsic(
           llvm_value, local_var, dbg_info_->di_builder_->createExpression(), llvm::DebugLoc(di_loc),
           llvm::BasicBlock::iterator(insert_before));
@@ -2213,7 +2214,7 @@ void CodeGenLLVM::AddDebugInformation(llvm::Value* llvm_value, const Var& tir_va
   }
 
   if (insert_before) {
-#if TVM_LLVM_VERSION >= 200
+#if TVM_LLVM_INSERTDECLARE_USES_ITERATOR
     dbg_info_->di_builder_->insertDeclare(
         llvm_value, local_var, dbg_info_->di_builder_->createExpression(), llvm::DebugLoc(di_loc),
         llvm::BasicBlock::iterator(insert_before));

